### DISCUSSION TITLE
Get entrypoint name in webpack loader

### DISCUSSION THREAD

**Discussion by nolimitdev** [webpack/webpack#14841] [General]
Im wondered that nowhere is answered how to access entry point name in webpack loader. Is it possible? For example in postcss.config.js are function arguments https://v4.webpack.js.org/loaders/postcss-loader/#examples-of-config-files api.file, api.mode, api.webpackLoaderContext, api.env, api.options but noone with entry name. It is useful to have little bit different configuration of webpack loader depending on entry point without having to create separete/duplicate webpack configuration file. 

On stackoverflow here https://stackoverflow.com/a/51604949/3826175 is the answer to use options.entry but it is obsolete and not working now. Another identical ask is here https://stackoverflow.com/questions/56149745/get-entrypoint-name-in-webpack-loader but with no answer for almost 3 years.

**Comment by alexander-akait** [comment]
Why you need this?

**Reply by nolimitdev** [reply]
I added use case here https://github.com/webpack-contrib/postcss-loader/issues/555 to webpack-loader feature request but Im afraid that webpack does not expose entry to loaders (but maybe yes).

> For example when we need some small configuration diff for postcss plugins e.g. for foo entry include some file for importFrom option of https://www.npmjs.com/package/postcss-custom-properties#user-content-importfrom and for entry bar include another file. The same example coud be for autoprefixer, for mixins etc.

**Reply by gottayan** [reply]
did you solve this?

**Reply by nolimitdev** [reply]
No and by answers here https://github.com/webpack-contrib/postcss-loader/issues/555#issuecomment-992578029 it seems to be technically impossible to do some different things between entry points e.g. load different plugins. Why, I do not know. It seems that solution is to hard copy webpack config for each entry :(

**Reply by alexander-akait** [reply]
What is use case, please provide example, I think you really do not need this, entry point it the same module as other modules, I can provide example how to achieve this, but I want to check you understand why you try to do it and provide example how to better solve it

**Reply by gottayan** [reply]
> What is use case, please provide example, I think you really do not need this, entry point it the same module as other modules, I can provide example how to achieve this, but I want to check you understand why you try to do it and provide example how to better solve it

You can look this.but I want to use the different loader to the different entry.
https://stackoverflow.com/questions/59944550/get-webpack-entry-point-name-and-reuse-it-in-loader.

**Reply by gottayan** [reply]
> What is use case, please provide example, I think you really do not need this, entry point it the same module as other modules, I can provide example how to achieve this, but I want to check you understand why you try to do it and provide example how to better solve it

to me，I just want to use postcss-plugin-px2rem in the mobile entry, not laptop.
for example:
<img width="724" alt="企业微信截图_34d808ed-2eac-42b0-9343-d9f7ef0fc0df" src="https://user-images.githubusercontent.com/23098167/145929482-522c9f8b-4826-4ee4-8df0-930cc62e2a7a.png">

**Reply by nolimitdev** [reply]
> What is use case, please provide example, I think you really do not need this, entry point it the same module as other modules, I can provide example how to achieve this, but I want to check you understand why you try to do it and provide example how to better solve it

I have provided use case here: https://github.com/webpack-contrib/postcss-loader/issues/555 and it is:
> For example when we need some small configuration diff for postcss plugins e.g. for foo entry include some file for `importFrom` option of https://www.npmjs.com/package/postcss-custom-properties#user-content-importfrom and for entry bar include another file. The same example could be for autoprefixer, for mixins etc.

I do not think that setting different `importFrom` of  `postcss-custom-properties` depending on entry could break something in webpack (loaded configuration is the same). It is plugin "side" and for webpack does not matter what file with CSS colors is used by plugin `postcss-custom-properties`. So it will be great to have this feature https://github.com/webpack-contrib/postcss-loader/issues/555 = official ability to know entry name which is being processed.

Exact example... Project/repository consist of 2 parts: admin part and user part (consider admin and user entry). Both parts must be compiled together because they use separate resources but also shared resources. When shared resource is changed admin + user must be recompiled but when recompiling correct CSS var file must be taken into account.

*webpack.config.js*
```
module.exports = (env, options) => {
    return {
        entry : {
            admin : './admin.js',
            user : './user.js',
        },
    }
};
```

*postcss.config.js*
```
// note that api.entryName is fake (feature request)
module.exports = (api) => {
    var cssColorsFile = (api.entryName == 'admin') ? './vars-admin.css' : './vars-user.css'
    
    return {
        plugins : [
            ['postcss-custom-properties', { importFrom : cssColorsFile }],
        ], 
    }
};
```

*admin.js*
```
import './global.css'; // shared resource for admin & user
import './admin.css'; // admin only resource
... etc JS code
```

*user.js*
```
import './global.css'; // shared resource for admin & user
import './user.css'; // user only resource
... etc JS code
```
So in this case it is not possible to check processed file name in `api.file`, because some files are part unique and some are shared. E.g. when `api.file == global.css` we do not know for which entry name.

UPDATE: it seems to be technically impossible, because when I console logged `api.file` in `postcss.config.js` I see that shared resource `global.css` is processed once (not twice for each entry so it not possible to decide which css colors file use). Is it right @alexander-akait ?

**Reply by alexander-akait** [reply]
> You can look this.but I want to use the different loader to the different entry.
https://stackoverflow.com/questions/59944550/get-webpack-entry-point-name-and-reuse-it-in-loader.

You can use `layer`, here small examples for polyfills https://github.com/webpack/webpack/discussions/13241#discussioncomment-674844, you can use the same strategy, look at the second example.

Anyway, each entry is just a module, even more you can import entry inside other entry and you can use `dependOn` to do the same on configuration level (and more you can do not have entries - module federation), so entry is pseudo abstraction, you have `api.file` and it is only thing which you can rely (and when you setup own entries you know their names). 

Don't forget when webpack build something, plugins can add more entries too (dynamic entries), so you can start to build a project with 1 entry and finish with 20 entries.

This is not implement not because I don't want to do it, because it protects you from invalid configurations and unexpected behavior and entries will be finished only when modules were built, if you really need to get entry you can use:
```
api.webpackLoaderContext._compilation.options.entry
// All webpack options are normalized and keeping here
// api.webpackLoaderContext._compilation.options
```

loader is just transformer and should not know anything extra, otherwise you will get a confusing structure and eventually get confused in it yourself

**Reply by nolimitdev** [reply]
1.) Here https://stackoverflow.com/questions/59944550/get-webpack-entry-point-name-and-reuse-it-in-loader is no solution

2.) But `api.webpackLoaderContext._compilation.options.entry` does not keep one/current entry but both from webpack.config.js - console.log of `api.webpackLoaderContext._compilation.options.entry`:
```
{
  admin: { import: [ './admin.js' ] },
  user: { import: [ './user.js' ] }
}
```

So as I mentioned it seems to be technically impossible in principle. For everyone the only one solution is to hard copy webpack configuration (disadvantage is that build time will be higher because shared resources will be compiled twice). To avoid hard copying/duplicating same code here https://stackoverflow.com/a/51604949/3826175 is interesting solution to use loop for webpack configuration and do some differences during loop.

**Reply by gottayan** [reply]
> 1.) Here https://stackoverflow.com/questions/59944550/get-webpack-entry-point-name-and-reuse-it-in-loader is no solution
> 
> 2.) But `api.webpackLoaderContext._compilation.options.entry` does not keep one/current entry but both from webpack.config.js - console.log of `api.webpackLoaderContext._compilation.options.entry`:
> 
> ```
> {
>   admin: { import: [ './admin.js' ] },
>   user: { import: [ './user.js' ] }
> }
> ```
> 
> So as I mentioned it seems to be technically impossible in principle. For everyone the only one solution is to hard copy webpack configuration (disadvantage is that build time will be higher because shared resources will be compiled twice). To avoid hard copying/duplicating same code here https://stackoverflow.com/a/51604949/3826175 is interesting solution to use loop for webpack configuration and do some differences during loop.

@alexander-akait Yes.I try to use layer but not working.also it can not solute these.

**Reply by alexander-akait** [reply]
> But api.webpackLoaderContext._compilation.options.entry does not keep one/current entry but both from webpack.config.js - console.log of api.webpackLoaderContext._compilation.options.entry:

Because we don't know current entry point, we just build modules, any entry points can be used in other entry points as I described above. 

> disadvantage is that build time will be higher because shared resources will be compiled twice

Where? If you two files - you need build two files.

Again, entry point is just abstraction, entry is a module, like other modules, they can be reused in many places. Even more - in your example you don't have `admin.css` or `user.css` entrypoints, I don’t know why you’re talking about them like that, because they are modules. How we can detect what is entry what is not?

**Comment by gottayan** [comment]
webpack5 do not support layer property.I try to this example but catch some error.

发自我的iPhone


------------------ Original ------------------
From: Alexander Akait ***@***.***&gt;
Date: Tue,Dec 14,2021 6:59 PM
To: webpack/webpack ***@***.***&gt;
Cc: gottayan ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [webpack/webpack] Get entrypoint name in webpack loader (Discussion #14841)





  
You can look this.but I want to use the different loader to the different entry.
 https://stackoverflow.com/questions/59944550/get-webpack-entry-point-name-and-reuse-it-in-loader.
  
You can use layer, here small examples for polyfills #13241 (comment), you can use the same strategy, look at the second example.
 
Anyway, each entry is just a module, even more you can import entry inside other entry and you can use dependOn to do the same on configuration level (and more you can do not have entries - module federation), so entry is pseudo abstraction, you have api.file and it is only thing which you can rely (and when you setup own entries you know their names).
 
Don't forget when webpack build something, plugins can add more entries too (dynamic entries), so you can start to build a project with 1 entry and finish with 20 entries.
 
This is not implement not because I don't want to do it, because it protects you from invalid configurations and unexpected behavior and entries will be finished only when modules were built, if you really need to get entry you can use:
 api.webpackLoaderContext._compilation.options.entry // All webpack options are normalized and keeping here // api.webpackLoaderContext._compilation.options  
loader is just transformer and should not know anything extra, otherwise you will get a confusing structure and eventually get confused in it yourself
 
—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by alexander-akait** [reply]
Please provide example how you do it, without code I can't say what is wrong, layers used in many places and there are a lot of tests on this feature

**Reply by gottayan** [reply]
> Please provide example how you do it, without code I can't say what is wrong, layers used in many places and there are a lot of tests on this feature

the same as the example you give: https://github.com/webpack/webpack/discussions/13241#discussioncomment-674844

// webpack.config.js
// ...
entry: {
    modern: {
      import: path.resolve(__dirname, './src/mobile/index.ts'),
      layer: "modern"
    },
    legacy: {
      import: path.resolve(__dirname, './src/laptop/index.ts'),
      layer: "legacy"
    },
  },
module.rules: [
{
        test: /\.(ts|js)x?$/,
        exclude: /node_modules/,
        oneOf: [{
          issuerLayer: "legacy",
          loader: "babel-loader",
          options: {
            useBuiltIns: "usage",
            browserslistEnv: "legacy"
          }
        }, {
          issuerLayer: "modern",
          loader: "babel-loader",
          options: {
            useBuiltIns: "usage",
            browserslistEnv: "modern"
          }
        }]
      },
]

Error: Compiling RuleSet failed: Properties issuerLayer are unknown (at clonedRuleSet-1[0].rules[0].oneOf[0]: [object Object])

**Reply by alexander-akait** [reply]
Do you enable `layer`s https://webpack.js.org/configuration/experiments/ in `experiments`?

**Reply by gottayan** [reply]
yeah，has set.

**Reply by nolimitdev** [reply]
The same error: `Error: Compiling RuleSet failed: Properties issuerLayer are unknown (at clonedRuleSet-2.oneOf[0]: [object Object])` with 

```
experiments : {
    layers : true,
},
```

**Reply by alexander-akait** [reply]
Can you provide full configuration, because we have https://github.com/webpack/webpack/blob/main/schemas/WebpackOptions.json#L4002

**Reply by nolimitdev** [reply]
@gottayan what is your error trace? Mine is...

```
[webpack-cli] Error: Compiling RuleSet failed: Properties issuerLayer are unknown (at clonedRuleSet-2.oneOf[0]: [object Object])
    at RuleSetCompiler.error (node_modules\webpack\lib\rules\RuleSetCompiler.js:373:10)
    at RuleSetCompiler.compileRule (node_modules\webpack\lib\rules\RuleSetCompiler.js:196:15)
    at node_modules\webpack\lib\rules\RuleSetCompiler.js:154:9
    at Array.map (<anonymous>)
    at RuleSetCompiler.compileRules (node_modules\webpack\lib\rules\RuleSetCompiler.js:153:16)
    at RuleSetCompiler.compileRule (node_modules\webpack\lib\rules\RuleSetCompiler.js:192:30)
    at cloneRule (node_modules\vue-loader\dist\pluginWebpack5.js:173:42)
    at node_modules\vue-loader\dist\pluginWebpack5.js:101:31
    at Array.map (<anonymous>)
    at VueLoaderPlugin.apply (node_modules\vue-loader\dist\pluginWebpack5.js:101:14)
```
I have noticed that problem is with vue-loader and its VueLoaderPlugin when I was creating example configuration with layers for @alexander-akait and example was working fine until I implemented VUE into example. Im using latest  "vue" (3.2.26) and "vue-loader" (17.0.0).

**Reply by alexander-akait** [reply]
oh, looks like vue-loader modify loaders and break configuration, I think will be great to open an issue in their repo

**Reply by nolimitdev** [reply]
It is already reported 6 months ago with no answer here https://github.com/vuejs/vue-loader/issues/1849 Vue-loader has 141 issues open, so I do not think it will be fixed at all :(

**Reply by nolimitdev** [reply]
Similar issue here https://github.com/vuejs/vue-loader/issues/1861 also with no answer from maintainers

**Reply by gottayan** [reply]
@nolimitdev diffrent from yours.
<img width="762" alt="企业微信截图_32c3b0be-c43a-4ca8-8b4a-73345ba3f67c" src="https://user-images.githubusercontent.com/23098167/146372647-8e4d7333-bd4d-4ca9-b4e0-652b30151b85.png">

**Reply by gottayan** [reply]
@alexander-akait just create a simple project to test and it is ok.I can not find the reason about it.
When I have time,will give you a project which have full configuration.
@nolimitdev it will better if you can provide.

**Reply by gottayan** [reply]
@alexander-akait assuming this is ok, the layer is still can not help me slove the problem.
I need to use multiple loaders in one situation.But 'webpack module rules oneOf' just support one loaders.
like this:
`module: {
    rules: [
      {
        test: /\.(le|c)ss$/,
        oneOf: [
          {
            issuerLayer: 'laptop',
            loader: 'css-loader',
          },
          // multiple loaders is not support
          {
            issuerLayer: 'mobile',
            use: ['css-loader', 'postcss-loader'],
          }
        ],
      }
    ]
  }`

**Reply by nolimitdev** [reply]
A have created minimalistic working demo for layers https://github.com/nolimitdev/webpack-layer-example
@gottayan btw in this example I have successfully used more loaders - so it is supported by webpack

```
{
    test : /\.css$/,
    oneOf : [
        {
            issuerLayer: 'blue',
            use : [
                { loader : MiniCssExtractPlugin.loader },
                { loader : 'css-loader', options : { url : false, sourceMap : false } },
                { loader : 'postcss-loader', options : { sourceMap : false, postcssOptions : { plugins : [['postcss-custom-properties', { preserve : false, importFrom : { customProperties: { '--color': 'blue' } } }]] } } },
            ],
        },
        {
            issuerLayer: 'gray',
            use : [
                { loader : MiniCssExtractPlugin.loader },
                { loader : 'css-loader', options : { url : false, sourceMap : false } },
                { loader : 'postcss-loader', options : { sourceMap : false, postcssOptions : { plugins : [['postcss-custom-properties', { preserve : false, importFrom : { customProperties: { '--color': 'gray' } } }]] } } },
            ],
        },
    ]
},
```
(sorry for long one-liner)

**Reply by alexander-akait** [reply]
@nolimitdev You can provide different context for `postcss.config.js` if you want to keep it into own configuration

**Reply by gottayan** [reply]
> alexander

just an example.we want use it to do another something we need.
@nolimitdev actually if I remove the vue-loader, it is ok for me.thank you so much.

**Reply by nolimitdev** [reply]
@gottayan from this working repo https://github.com/nolimitdev/webpack-layer-example you do not need to remove vue-loader (this working repo is without it)

**Comment by gottayan** [comment]
ok.Thank you.



发自我的iPhone


------------------ Original ------------------
From: nolimitdev ***@***.***&gt;
Date: Thu,Dec 16,2021 8:53 PM
To: webpack/webpack ***@***.***&gt;
Cc: gottayan ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [webpack/webpack] Get entrypoint name in webpack loader(Discussion #14841)

**Comment by gottayan** [comment]
yeah.I do it in my another project.



发自我的iPhone


------------------ Original ------------------
From: nolimitdev ***@***.***&gt;
Date: Fri,Dec 17,2021 2:26 PM
To: webpack/webpack ***@***.***&gt;
Cc: gottayan ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [webpack/webpack] Get entrypoint name in webpack loader (Discussion #14841)





 
@gottayan from this working repo https://github.com/nolimitdev/webpack-layer-example you do not need to remove vue-loader (this working repo is without it)
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
